### ISSUE TITLE
Compilation errors in codyco-superbuild with eigen

### ISSUE THREAD

**Issue by rlober** [ocra-recipes/ocra-wbi-plugins#18]
I am porting an email chain to this issue:

# ISSUE 1

> Hi Guys,
> 
> As Vincent mentioned in his email this morning, we (at INRIA) plan to use the OCRA ISIR controller (probably Joseph's version, I think it is "wocra"). I installed ocra-recipes following the recommendations on github, it works fine. However, when I try to use it, I have issues with TinyXML.
> 
> In order to use OCRA, I put "find_package(OcraRecipes REQUIRED)" in the CMakeLists.txt of my project. I'm not even sure it's the way I should do, so please correct me if I'm wrong.
> 
> Assuming I did the correct thing to use OCRA, when I try to compile my project, I get the following error:
> 
> CMake Error at /usr/local/lib/cmake/OcraRecipes/OcraRecipesConfig.cmake:25 (find_package):
>   By not providing "FindTinyXML.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "TinyXML", but
>   CMake did not find one.
> 
>   Could not find a package configuration file provided by "TinyXML" with any
>   of the following names:
> 
>     TinyXMLConfig.cmake
>     tinyxml-config.cmake
> 
>   Add the installation prefix of "TinyXML" to CMAKE_PREFIX_PATH or set
>   "TinyXML_DIR" to a directory containing one of the above files. If
>   "TinyXML" provides a separate development package or SDK, be sure it has
>   been installed.
> 
> Yet, I followed all the instruction on the github page. In particular I installed libtinyxml-dev.
> 
> I don't know if this is me doing something wrong (I'm not really gifted with compilation ^^), or is there is an issue that is not explained on the github. I am using Ubuntu 16.04.
> 
> 
> Otherwise, I tried to install OCRA through the codyco-superbuild by setting the "OCRA" option ON, but I have compilation errors. So I dropped that and installed OCRA directly (which is when I have the problem above).
> 
> Thanks in advance for any help,
> 
> Pauline

# ISSUE 2

> Hi again,
> 
> Actually I think I more or less figure out how to use OCRA in my project (though some clarifications on my previous question would be appreciated), but I have 2 other questions.
> 
> The 1st one is about ocra-wbi-plugins. I'd like to install it, but when I try with the codyco-superbuild (as suggested on github) by setting the CODYCO_BUILD_OCRA_MODULES option ON, I have compilation errors, which seemed related to Eigen. I'm wondering whether this is a "normal" current issue (and if there is a fix to it), or if I'm just doing something wrong (though I followed all the codyco-superbuild instructions). In particular, can you tell me what Eigen version I should use? The codyco-superbuild page says that Eigen 3.3 is required, but the ocra page says it's 3.2, so I'm lost...
> 
> The 2nd question is more applied. Have you used the whole controller/simulation with a robot other than iCub? Basically, I'm wondering whether there is somewhere in the codyco packages a tool to import a URDF and build a model suitable for ocra controllers from the URDF. I saw that ocra-wbi-plugins implement an interface class for the "Model" abstract class from ocra, based on WholeBodyInterface. But it is unclear to me if WBI works only for iCub, or for any kind of models. Specifically, it seems that WBI is only abstract classes, so I'm wondering whether there is a tool to use interface it with URDFs or whether I have to do the job myself! (just trying to avoid re-doing something that was already done).
> 
> Thanks a lot for your help,
> 
> Pauline

**Comment by rlober** [comment]
Hey Pauline,

1.  @jeljaik, is testing a fresh install on OSX to try to reproduce the problem. He will let you know what happens. 
1.  Theoretically you should be able to use WBI with any URDF. Jorh and I have just been using that of the iCub and aren't really specialists, so I would ask somebody like:
@traversaro 
@francesco-romano
@gabrielenava
@DanielePucci

**Comment by traversaro** [comment]
The current `ocra-wbi-plugins` and the WBI despite the original design goal, are actually quite YARP-dependent. 
Depending on the kind of robot you are interested in controlling, I would suggest to directly implement the necessary OCRA interfaces, similarly to what you have done with the Kuka LWR . If necessary, the model part can be implemented using the iDynTree class `iDynTree::KinematicsDynamicsComputations` .

@diegoferigo

**Comment by rlober** [comment]
CC: @paulinejmaurice

**Comment by rlober** [comment]
@traversaro 

I believe yarp dependencies aren't too problematic for the application. Not to mention that ocra is largely yarp dependent as well. 

Are there any guides for setting up a config file for a robot using yarpWBI?

**Comment by paulinejmaurice** [comment]
The robot I am trying to control is a URDF description of a human model, so I'm only doing simulations (with Gazebo, no "real" robot). I am not against using YARP-dependent modules, though I am new to YARP. I am currently looking at iDynTree for the model, but was wondering whether the yarpWholeBodyInterface could be used easily with URDF models.

**Comment by jeljaik** [comment]
Hello @paulinejmaurice 

We still need a few days before we can finally merge a version of OCRA without this compatibility issue (and without Eigen LGSM which is the source of all evil here). If you don't wanna wait too much though, and wanna start exploring the inner workings, etc, what you can do at this point is the following: 

1. Update `ocra-recipes` (assuming you're still in the `master` branch.
    ```
    cd $CODYCO_SUPERBUILD_ROOT
    cd libraries/ocra-recipes
    git pull
    ```
1. Remove completely your current build of `codyco-superbuild` 
    ```
    cd $CODYCO_SUPERBUILD_DIR
    rm -r *
    ```
2. Install a version of Eigen < 3.2.9 (But don't remove 3.3.4 needed by codyco-superbuild). Move to wherever you download code and do the following:
   ```
    git clone https://github.com/RLovelett/eigen
    git checkout tags/ae3b232588719cabb878c9bada44de3fdd77a9f8
    cd eigen
    mkdir build && cd build
    cmake ../ && make install
   ```
3. Try to install everything again
   ```
    cd $CODYCO_SUPERBUILD_DIR
    cmake ../ -DCODYCO_BUILD_OCRA_MODULES=ON
    make
   ```
4. Let us know what happens.    

I performed these steps with `gcc` on a MacBook with `macOS Sierra 10.12.5`.

**Comment by paulinejmaurice** [comment]
Thanks @jeljaik 
I tried to follow the above steps but still have issues.

1- The checkout for Eigen doesn't seem to work. But I installed Eigen 3.2.8 anyway, so I hope this should not be the issue.

2- When I try to install codyco-superbuild with the OCRA_MODULES option on, there is an error with ocra-wbi-plugins. The building of ocra-recipes is fine, but with ocra-wbi-plugins I get this error:
 "Performing download step (git clone) for 'ocra-wbi-plugins'
CMake Error at /home/pmaurice/codyco-superbuild/build/main/ocra-wbi-plugins/CMakeFiles/YCMTmp/ocra-wbi-plugins-gitclone.cmake:26 (message):
  "/home/pmaurice/codyco-superbuild/main/ocra-wbi-plugins" folder exists and
  is not a git repository.  Remove it and try again"

I tried removing the folder as suggested, but I get the same error again when I re-run "make".


So what I tried was to install codyco-superbuild without OCRA_MODULES, and then separately install ocra-recipes following the instructions from here https://github.com/ocra-recipes/ocra-recipes.
I don't know if that is a viable option (installing the two separately), so if not please tell me.

When I install ocra-recipes separately (with Eigen 3.2.8 installed), then the installation is OK.
However I have issues when I try to use it in my project. I guessed that I have to add "find_package(OcraRecipes REQUIRED)" in my CmakeLists file. The package seems to exist, but then I have the issue with TinyXML:

CMake Error at /usr/local/lib/cmake/OcraRecipes/OcraRecipesConfig.cmake:25 (find_package):
  By not providing "FindTinyXML.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "TinyXML", but
  CMake did not find one.

  Could not find a package configuration file provided by "TinyXML" with any
  of the following names:

    TinyXMLConfig.cmake
    tinyxml-config.cmake

  Add the installation prefix of "TinyXML" to CMAKE_PREFIX_PATH or set
  "TinyXML_DIR" to a directory containing one of the above files.  If
  "TinyXML" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:18 (find_package)

So I don't know if I am making a mistake somewhere, or if there is still an ocra related issue. Any help would be welcome!
Thanks

**Comment by traversaro** [comment]
Hi everyone! 
I think this is an `ocra-recipes` issue, in particular related on how `OcraRecipesConfig.cmake`. The problem is that `FindTinyXML.cmake` is not provided by CMake, so ocra-recipes CMake should take care of handling the case in which the downstream project does not have any `FindTinyXML.cmake` file in `CMAKE_MODULE_PATH` . 
Without fixing the issue in `ocra-recipes`, possible workarounds include: 
* Call `find_package(YCM REQUIRED)` before `find_package(OcraRecipes REQUIRED)` this should find the [`YCM`](https://github.com/robotology/ycm) installed in the superbuild that should add its own `FindTinyXML.cmake` to `CMAKE_MODULE_PATH`. 
* Add `FindTinyXML.cmake` in your project, for example following the structure of the `ocra-recipes` itself.

**Comment by rlober** [comment]
deprecated - cleanup
